contrib_bots/lib/gmail.py

```diff
@@ -0,0 +1,153 @@
+import mimetypes
+import os
+import re
+
+from googleapiclient import discovery, errors
+from email.mime.text import MIMEText
+from oauth2client.file import Storage
+
+class GmailHandler(object):
+    '''
+    This plugin facilitates sending emails from a Gmail account
+    from within Zulip. It looks for messages starting with '@gmail'.
+    '''
+
+    def usage(self):
+        return '''
+            This plugin will allow users to send emails from a
+            personal gmail account. Users should preface
+            messages with "@gmail".
+
+            Enter messages in the following format:
+            "@gmail --recipient <recipient_email> --subject <subject> --label <label> --body
+            <body>". The --label tag are optional and can be left out.
+
+            Before running this, make sure to create a bot and store its information
+            in ~/.zuliprc.
+            '''
+
+    def triage_message(self, message, client):
+        # return True if we want to (possibly) response to this message
+
+        original_content = message['content']
+        self.bot_email = client.email
+
+        # This next line of code is defensive
+        if message['display_recipient'] == 'gmail':
+            return False
+        return original_content.startswith('@gmail')
+
+    def handle_message(self, message, client, state_handler):
+        original_content = message['content']
+        original_sender = message['sender_email']
+
+        recipient, subject, label, body = parse_input(original_content)
+
+        service = get_service()
+        email = create_message(body, subject, recipient)
+        result = send_message(service, email)
+
+        label_result = -1
+        if label:
+            create_label(service, label)
+            label_id = get_label_id(service, label)
+            label_result = add_label(service, label_id, result['id'])
+
+        if 'SENT' in result['labelIds']:
+            content = 'Email sent to {}'.format(recipient)
+            if label_result:  # Enters this block if the label couldn't be added
+                              # or if there was no label
+                content += ' without label.'
+            else:
+                content += '.'
+        else:
+            content = 'Could not send email to {}. Please try again.'.format(recipient)
+
+        client.send_message(dict(
+            type = 'private',
+            to = original_sender,
+            sender = self.bot_email,
+            content = content
+        ))
+
+def parse_input(user_input):
+    # Input is of the following format: (--label flag is optional)
+    # @gmail --recipient <recipient_email> --subject <subject> --label <label> --body <body>
+    if '--label' not in user_input:
+        loc = user_input.index('--body')
+        user_input = user_input[:loc] + '--label ' + user_input[loc:]
+
+    regex = "--recipient (.*?) --subject (.*?) --label (.*?) --body (.*?)\Z"
+    return re.findall(regex, user_input)[0]
```

Comment:
You can simply use `re.match`; it will just search for one match at the start of the string.